streaming/base/util.py

```diff
@@ -196,3 +204,138 @@ def get_import_exception_message(package_name: str, extra_deps: str) -> str:
     return f'Streaming was installed without {package_name} support. ' + \
             f'To use {package_name} related packages with Streaming, run ' + \
             f'`pip install \'mosaicml-streaming[{package_name}]\'`.'
+
+
+def merge_index(folder_urls: List[Union[str, Tuple[str, str]]],
+                out: Union[str, Tuple[str, str]],
+                *,
+                keep_local: bool = True,
+                overwrite: bool = True,
+                download_timeout: int = 100) -> int:
+    """Merge index.json from a list of remote or local directories.
+
+    Args:
+        folder_urls (Iterable): folders that contain index.json for the partition
+            each element can take the form of a single path string or a tuple string
+
+            for each  url in folder_urls, if url is
+                1. tuple (local, remote): check if local is accessible.
+                    -> Yes: use local index to merge
+                    -> No:  download from remote first, then merge
+                2. str (local path): use local path to merge.
+                    raise FileNotFoundError if any local index is not accessible
+                3. str (remote url): download to a temp directory first, then merge
+
+        out (Union[str, Tuple[str, str]]): path to put the merged index file
```

Comment:
Wondering, why we do need `out` argument in this method? Generally, for merging index, user would provide either `merge_index(local_dir, remote_dir)` path or `merge_index(local_dir)` or `merge_index(remote_dir)`. Depending on which use-case out of above all three, the final merged index.json would either store in local_dir or remote_dir or both. We don't expect the final merged index.json to be saved else apart from above three location.

Reply:
Will keep this method as a helper as it is convenient for dataframeToMDS. Adding a wrapper merge_index function with just a str path as argument.